File: java/exercises/practice/unit_test_474c8b9b_7a30_411a_bfbd_12c33bf235fa/rate_limiter/.meta/src/main/java/RateLimiter.java
import java.util.concurrent.ConcurrentHashMap;

public class RateLimiter {

    private volatile int maxRequests;
    private volatile long windowMillis;
    private final ConcurrentHashMap<String, UserRecord> userRecords;

    public RateLimiter(int maxRequests, long windowMillis) {
        this.maxRequests = maxRequests;
        this.windowMillis = windowMillis;
        this.userRecords = new ConcurrentHashMap<>();
    }

    /**
     * Allows a request for the given user if it is within the rate limit.
     * @param user a unique identifier for a user (can be user ID, IP, etc.)
     * @return true if the request is allowed, false otherwise.
     */
    public boolean allowRequest(String user) {
        long now = System.currentTimeMillis();
        // Create or get the current record for the user.
        UserRecord record = userRecords.computeIfAbsent(user, k -> new UserRecord(now, 0));

        // Synchronize on the record to ensure that checking and updating is atomic for that user.
        synchronized (record) {
            // If the current window has expired. Reset the record.
            if (now - record.windowStart >= windowMillis) {
                record.windowStart = now;
                record.count = 1;
                return true;
            } else {
                // Within current window. If request count is under limit, increment.
                if (record.count < maxRequests) {
                    record.count++;
                    return true;
                } else {
                    return false;
                }
            }
        }
    }

    /**
     * Dynamically update the rate limit configuration.
     * @param newMaxRequests new maximum number of requests allowed in the specified window.
     * @param newWindowMillis new time window in milliseconds.
     */
    public void updateLimit(int newMaxRequests, long newWindowMillis) {
        this.maxRequests = newMaxRequests;
        this.windowMillis = newWindowMillis;
    }

    /**
     * Inner class to represent a user's request record.
     */
    private static class UserRecord {
        long windowStart;
        int count;

        UserRecord(long windowStart, int count) {
            this.windowStart = windowStart;
            this.count = count;
        }
    }
}

File: javascript/exercises/practice/unit_test_643963c4_8724_4840_841c_2973ece6ecb3/rate_limiter/.meta/rate_limiter.js
'use strict';

// Default global configuration values
const DEFAULT_REQUEST_LIMIT = 100;
const DEFAULT_TIME_WINDOW = 60; // in seconds

// Use the global user configuration store if available; otherwise, create a new Map.
const userLimits = global.userLimits || new Map();

/**
 * Configure a specific rate limit for a user. This configuration overrides the default limit.
 *
 * @param {string} userId - Unique identifier for the user (alphanumeric characters only).
 * @param {number} requestLimit - Maximum allowed number of requests within the time window.
 * @param {number} timeWindow - Time window in seconds.
 * @throws Will throw an error if input parameters are invalid.
 */
function configureUserLimit(userId, requestLimit, timeWindow) {
  if (typeof userId !== 'string' || !/^[a-z0-9]+$/i.test(userId)) {
    throw new Error('Invalid userId');
  }
  if (!Number.isInteger(requestLimit) || requestLimit <= 0) {
    throw new Error('requestLimit must be a positive integer');
  }
  if (!Number.isInteger(timeWindow) || timeWindow <= 0) {
    throw new Error('timeWindow must be a positive integer');
  }
  userLimits.set(userId, { requestLimit, timeWindow });
}

/**
 * Determine if a request from a given user is allowed based on the configured rate limit.
 *
 * The function calculates the current time window based on the user's configuration (or default configuration)
 * and uses an atomic increment operation to count the number of requests within the current window.
 * If the count exceeds the request limit, the function returns false, indicating that the request is blocked.
 *
 * @param {string} userId - Unique identifier for the user.
 * @returns {boolean} - Returns true if the request is allowed; false otherwise.
 */
function isAllowed(userId) {
  // Get the configuration for this user, or fall back to default configuration.
  let config = userLimits.get(userId);
  if (!config) {
    config = { requestLimit: DEFAULT_REQUEST_LIMIT, timeWindow: DEFAULT_TIME_WINDOW };
  }
  const { requestLimit, timeWindow } = config;

  // Get the current time (in seconds). Assumes a global getTimeInSeconds() function.
  const currentTime = global.getTimeInSeconds();
  // Calculate the current time window start time.
  const windowStart = Math.floor(currentTime / timeWindow) * timeWindow;
  
  // Construct a unique key for this user's current time window.
  const key = `rl:${userId}:${windowStart}`;
  
  // Atomically increment the request count for the given key. The key expires after timeWindow seconds.
  // Assumes a global atomicIncrement(key, expiryInSeconds) function.
  const requestCount = global.atomicIncrement(key, timeWindow);
  
  // If the number of requests does not exceed the limit, the request is allowed.
  return requestCount <= requestLimit;
}

module.exports = {
  isAllowed,
  configureUserLimit
};